File: desktop/src/lib/chatFormat.ts
import type { ArchcarProjectionItem } from "@/bridge/protocol";

// GTK chat-surface parity: which projected render classes we actually render and
// how. Ports session_surface.rs's provider_projection_item_widget dispatch — the
// timeline is a strict allowlist, everything else is dropped rather than shown as
// a raw blob.

// Prose rendered as markdown (chat_text_markup in GTK).
export const TEXT_CLASSES = new Set(["user_chat", "assistant_chat", "reasoning_card"]);

// Inline "chip-card" event classes → the fallback verb when the title carries no
// action word. Commands "Ran", tools "Used", file reads "Read", diffs "Edited".
// Any class not listed here (and not a text class) is ditched.
const DEFAULT_VERB: Record<string, string> = {
  command_card: "Ran",
  process_card: "Ran",
  background_card: "Ran",
  file_card: "Read",
  diff_card: "Edited",
  skill_card: "Used",
  tool_card: "Used",
  plugin_card: "Used",
  subagent_card: "Used",
  nested_transcript_card: "Used",
};

// GTK's action-label prefixes: when the projected title already starts with one
// of these verbs, we keep it as the action and use the rest as the chip so we
// don't double up (e.g. "Ran ls" → verb "Ran", chip "ls").
const ACTION_PREFIXES = new Set([
  "Ran",
  "Read",
  "Used",
  "Opened",
  "Added",
  "Edited",
  "Deleted",
  "Searched",
]);

export function isInlineEventClass(renderClass: string): boolean {
  return renderClass in DEFAULT_VERB;
}

/** True if we render this item at all — a strict allowlist (text or inline event). */
export function isRenderableClass(renderClass: string): boolean {
  return TEXT_CLASSES.has(renderClass) || isInlineEventClass(renderClass);
}

/**
 * Split an inline event into its action verb and content chip, mirroring GTK's
 * inline_event_action_label: if the title leads with a known verb, that's the
 * action and the remainder is the chip; otherwise fall back to a per-class verb
 * (tools → "Used", commands → "Ran", …) with the whole title as the chip.
 */
export function inlineEventVerbChip(renderClass: string, title: string): { verb: string; chip: string } {
  const trimmed = title.trim();
  const space = trimmed.indexOf(" ");
  const first = space === -1 ? trimmed : trimmed.slice(0, space);
  if (ACTION_PREFIXES.has(first)) {
    return { verb: first, chip: trimmed.slice(space + 1).trim() };
  }
  return { verb: DEFAULT_VERB[renderClass] ?? "Used", chip: trimmed };
}

// Archductor-internal blocks injected into prompts/responses that must not be
// shown to the user (port of core's strip_archductor_metadata_block +
// hidden-instruction/attachment stripping, applied at render time).
const BLOCK_TAGS = ["archductor_metadata", "archductor_hidden_instruction"];

export function stripArchductorMetadata(text: string): string {
  let out = text;
  for (const tag of BLOCK_TAGS) {
    out = out.replace(new RegExp(`<${tag}>[\\s\\S]*?</${tag}>`, "g"), "");
  }
  // Self-closing attachment references (<archductor_attachment ... />).
  out = out.replace(/<archductor_attachment\b[^>]*\/>/g, "");
  return out.replace(/\n{3,}/g, "\n\n").trim();
}

export function isDiffCard(item: ArchcarProjectionItem): boolean {
  return item.render_class === "diff_card";
}

export function isTerminalCard(item: ArchcarProjectionItem): boolean {
  return (
    item.render_class === "command_card" ||
    item.render_class === "process_card" ||
    item.render_class === "background_card"
  );
}

File: ios/ArchcarKit/Sources/ArchcarKit/Stores/ChatFormat.swift
import Foundation

/// Render-time cleanup of agent text.
///
/// Archductor injects blocks into prompts and responses for its own use — the
/// summary/naming metadata, hidden instructions, attachment references. They
/// are part of the transcript but not part of the conversation, and core keeps
/// them in the projection body, so every surface strips them at render time.
/// This is a port of `desktop/src/lib/chatFormat.ts`, which in turn ports
/// core's `strip_archductor_metadata_block`; a surface that skips it shows the
/// user a JSON blob above every reply.
public enum ChatFormat {
    private static let blockTags = ["archductor_metadata", "archductor_hidden_instruction"]

    /// Prose classes, rendered as text.
    private static let textClasses: Set<String> = ["user_chat", "assistant_chat", "reasoning_card"]

    /// Inline activity classes and the verb to use when the title carries no
    /// action word of its own.
    private static let defaultVerb: [String: String] = [
        "command_card": "Ran",
        "process_card": "Ran",
        "background_card": "Ran",
        "error_card": "Error",
        "file_card": "Read",
        "diff_card": "Edited",
        "skill_card": "Used",
        "tool_card": "Used",
        "plugin_card": "Used",
        "subagent_card": "Used",
        "nested_transcript_card": "Used"
    ]

    private static let actionPrefixes: Set<String> = [
        "Ran", "Read", "Used", "Opened", "Added", "Edited", "Deleted", "Searched"
    ]

    /// Assistant prose renders only once finalized: partial markdown churns
    /// while it streams. Activity cards render live so the agent's current
    /// state is visible.
    private static let textFinalizedOnly: Set<String> = ["assistant_chat"]

    /// A strict allowlist, ported from the desktop's `isRenderableClass`.
    /// Anything else — hook cards, mcp chatter, classes core adds later — is
    /// dropped rather than shown as a raw event. A Claude session emits six
    /// hook cards at startup, which on a phone would be the whole first screen.
    public static func isRenderable(renderClass: String) -> Bool {
        textClasses.contains(renderClass) || defaultVerb[renderClass] != nil
    }

    public static func isDisplayable(_ item: ProjectionItem) -> Bool {
        guard isRenderable(renderClass: item.renderClass) else { return false }
        if textFinalizedOnly.contains(item.renderClass) { return item.streamState == "complete" }
        return true
    }

    /// Splits an activity card's title into an action verb and a content chip,
    /// so "Ran cargo test" reads as a verb plus what it acted on instead of
    /// doubling the verb up.
    public static func verbChip(renderClass: String, title: String) -> (verb: String, chip: String) {
        let trimmed = title.trimmingCharacters(in: .whitespaces)
        let first = trimmed.split(separator: " ", maxSplits: 1).first.map(String.init) ?? trimmed
        if actionPrefixes.contains(first) {
            let rest = trimmed.dropFirst(first.count).trimmingCharacters(in: .whitespaces)
            return (first, rest)
        }
        return (defaultVerb[renderClass] ?? "Used", trimmed)
    }

    public static func displayText(_ text: String) -> String {
        var out = text
        for tag in blockTags {
            out = out.replacingOccurrences(
                of: "<\(tag)>[\\s\\S]*?</\(tag)>", with: "", options: .regularExpression)
        }
        // Self-closing attachment references.
        out = out.replacingOccurrences(
            of: "<archductor_attachment\\b[^>]*/>", with: "", options: .regularExpression)
        return out
            .replacingOccurrences(of: "\n{3,}", with: "\n\n", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
